core/src/main/scala/kafka/admin/ReassignPartitionsCommand.scala

```diff
@@ -196,26 +431,33 @@ object ReassignPartitionsCommand extends Logging {
     (partitionsToBeReassigned, currentAssignment)
   }
 
-  def executeAssignment(zkClient: KafkaZkClient, adminClientOpt: Option[Admin], opts: ReassignPartitionsCommandOptions): Unit = {
+  def generateAssignment(zkClient: KafkaZkClient, brokerListToReassign: Seq[Int], topicsToMoveJsonString: String,
+                         disableRackAware: Boolean): (Map[TopicPartition, Seq[Int]], Map[TopicPartition, Seq[Int]]) = {
+    val serviceClient = ZkClientReassignCommandService(zkClient)
+    generateAssignment(serviceClient, brokerListToReassign, topicsToMoveJsonString, disableRackAware)
+
+  }
+
+  def executeAssignment(serviceClient : ReassignCommandService, adminClientOpt: Option[Admin], opts: ReassignPartitionsCommandOptions): Unit = {
     val reassignmentJsonFile =  opts.options.valueOf(opts.reassignmentJsonFileOpt)
     val reassignmentJsonString = Utils.readFileAsString(reassignmentJsonFile)
     val interBrokerThrottle = opts.options.valueOf(opts.interBrokerThrottleOpt)
     val replicaAlterLogDirsThrottle = opts.options.valueOf(opts.replicaAlterLogDirsThrottleOpt)
     val timeoutMs = opts.options.valueOf(opts.timeoutOpt)
-    executeAssignment(zkClient, adminClientOpt, reassignmentJsonString, Throttle(interBrokerThrottle, replicaAlterLogDirsThrottle), timeoutMs)
+    executeAssignment(serviceClient, adminClientOpt, reassignmentJsonString, Throttle(interBrokerThrottle, replicaAlterLogDirsThrottle), timeoutMs)
   }
 
-  def executeAssignment(zkClient: KafkaZkClient, adminClientOpt: Option[Admin], reassignmentJsonString: String, throttle: Throttle, timeoutMs: Long = 10000L): Unit = {
-    val (partitionAssignment, replicaAssignment) = parseAndValidate(zkClient, reassignmentJsonString)
-    val adminZkClient = new AdminZkClient(zkClient)
-    val reassignPartitionsCommand = new ReassignPartitionsCommand(zkClient, adminClientOpt, partitionAssignment.toMap, replicaAssignment, adminZkClient)
+  def executeAssignment(serviceClient : ReassignCommandService, adminClientOpt: Option[Admin],
+                        reassignmentJsonString: String, throttle: Throttle, timeoutMs: Long): Unit = {
+    val (partitionAssignment, replicaAssignment) = parseAndValidate(serviceClient, reassignmentJsonString)
+    val reassignPartitionsCommand = new ReassignPartitionsCommand(serviceClient, partitionAssignment.toMap, replicaAssignment)
 
     // If there is an existing rebalance running, attempt to change its throttle
-    if (zkClient.reassignPartitionsInProgress()) {
+    if (serviceClient.reassignInProgress) {
```

Comment:
I'm not sure we want to limit this when we have a reassignment in progress. The goal of KIP-455 is to allow incremental reassignments